### PR TITLE
Use newtype names as the underlying variable names

### PR DESCRIPTION
**Pull request by mrmr1993**
This is fairly self-explanatory: if the name of a `Pexp_newtype` meets the criteria for a type variable name (checks associated with the error `Invalid_variable_name`), then we use that name for the variable we generate.

### CODE REVIEW

**Comment by hhugo** [comment]
Could you add a test with `(type a')`

**Comment by mrmr1993** [comment]
> Could you add a test with `(type a')`

Done, thanks!

**Comment by garrigue** [comment]
This does make sense.
Note that the impact is small, because `type a b c. <typexp>` is processed at the syntactic level before expanding to `(type a b c)`, and generates a type annotation with with the same variable names.

File: typing/typecore.ml [hunk 1]
```diff
@@ -3230,7 +3230,12 @@ and type_expect_
       re { exp with exp_extra =
              (Texp_poly cty, loc, sexp.pexp_attributes) :: exp.exp_extra }
   | Pexp_newtype({txt=name}, sbody) ->
-      let ty = newvar () in
+      let ty =
+        if name <> "" && name.[0] = '_' then (* Invalid variable name. *)
```
Comment (alainfrisch) [inline typing/typecore.ml]: `Typetexp` (or Types,....?) could expose a `val valid_tyvar_name: string -> bool` (and use it internally in transl_type_param).

**Comment by mrmr1993** [comment]
Done! I changed the definition of what a valid name is slightly; I assume that the empty string shouldn't be a valid name..?

**Comment by mrmr1993** [comment]
@garrigue do you think this will make the next version? I've been suffering with the traditional generated names when module signatures don't match, it would help me a lot to have the `(type ...)`d names!

**Comment by trefis** [comment]
@mrmr1993 if by next version you mean 4.08, for which we've had a couple of beta releases already, then I believe the answer is no.

**Comment by garrigue** [comment]
It will be merged in trunk, which means being available for the next release, which will be 4.09.
You're free to use trunk (or a patched version of the compiler) if you need this earlier.

Just now I'm waiting for the CI check. It was failing (due to a conflict?).

File: typing/typecore.ml [hunk 2]
```diff
@@ -3238,7 +3238,12 @@ and type_expect_
       re { exp with exp_extra =
              (Texp_poly cty, loc, sexp.pexp_attributes) :: exp.exp_extra }
   | Pexp_newtype({txt=name}, sbody) ->
-      let ty = newvar () in
+      let ty =
+        if Typetexp.valid_tyvar_name name then
+          newvar ()
+        else
+          newvar ~name ()
+      in
```
Comment (garrigue) [inline typing/typecore.ml:3246]: It looks like it is the other way round. Shouldn't you use `name` only if it is valid?

Comment (garrigue) [reply typing/typecore.ml:3246]: This seems to be the reason the CI fails.

Comment (mrmr1993) [reply typing/typecore.ml:3246]: Woops, sorry! Should be fixed now.

**Comment by garrigue** [comment]
Thank you!

**Pull request merged**